provider/metrics/container_insights.go

```diff
@@ -0,0 +1,315 @@
+package metrics
+
+import (
+	"context"
+	"time"
+
+	"github.com/patrickmn/go-cache"
+	"github.com/pkg/errors"
+	"github.com/thoas/go-funk"
+	"github.com/virtual-kubelet/azure-aci/client/aci"
+	"github.com/virtual-kubelet/virtual-kubelet/log"
+	stats "github.com/virtual-kubelet/virtual-kubelet/node/api/statsv1alpha1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// container insights implementation of podStatsGetter interface
+type containerInsightsPodStatsGetter struct {
+	metricsGetter        ContainerGroupMetricsGetter
+	resourceGroup        string
+	cumulativeUsageCache *cache.Cache
+}
+
+// some of values in stats.PodStats are require cumulative data. But Container Insights be only able to
+// provide the average value in 1 minute period. So we have to calculate the cumulative value from the average.
+// we need to remember the last cumulative value and add new value of next time windows.
+type podCumulativeUsage struct {
+	podUID                         string
+	containersUsageCoreNanoSeconds map[string]cumulativeUsage
+	networkRx                      cumulativeUsage
+	networkTx                      cumulativeUsage
+}
+
+type cumulativeUsage struct {
+	lastUpdateTime time.Time
+	value          uint64
+}
+
+func NewContainerInsightsMetricsProvider(metricsGetter ContainerGroupMetricsGetter, resourceGroup string) *containerInsightsPodStatsGetter {
+	return &containerInsightsPodStatsGetter{
+		metricsGetter:        metricsGetter,
+		resourceGroup:        resourceGroup,
+		cumulativeUsageCache: cache.New(time.Minute*30, time.Minute*30),
```

Comment:
Ditto on LRU cache.

Reply:
Here is the second scenarios.  We cache the value for calculating cumulative value. we don't expect the cache be clean if cache run out of capacity.

Reply:
Maybe some docs on the cumulative value.
Are you saying the value can't be recalculated?

Reply:
for example the cumulative value for CPU usage. the value should be totally CPU usage from container start up.  If we want to recalculated it, we need query the metrics data start from time point when container created every time. To avoid do that, we cached the value and just add the new value on top of that.